### PR TITLE
feature:  Warn user if they leave the /editor page with unsaved changes #109

### PR DESCRIPTION
**Pull request by Satvik1769**
This pr solves issue https://github.com/tscircuit/snippets/issues/109
Using beforeUnload and popStatus event I solved the issue and tested it in various ways manually.
I believe conflicts will not occur now
Please merge this pr as soon as possible.

Regards,
Satvik

### CODE REVIEW

**Comment by seveibar** [comment]
@Satvik1769 failing format check, you can fix with `bun format`

File: src/components/CodeEditor.tsx [hunk 1]
```diff
@@ -221,7 +227,7 @@ export const CodeEditor = ({
             hoverTooltip((view, pos, side) => {
               const { from, to, text } = view.state.doc.lineAt(pos)
               const line = text.slice(from, to)
-              const match = line.match(/@tsci\/[\w\-.]+/)
+              const match = line.match(/@tsci\/[\w.]+/)
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:230]: you reverted some fixes here. Make sure to only commit changes that are relevent for your fix

File: src/components/CodeEditor.tsx [hunk 2]
```diff
@@ -331,6 +337,43 @@ export const CodeEditor = ({
...
+    if (hasUnsavedChanges) {
+      window.history.pushState(null, "", window.location.href)
+    }
+
+    // Cleanup event listeners on component unmount
+    return () => {
+      window.removeEventListener("beforeunload", handleBeforeUnload)
+      window.removeEventListener("popstate", handlePopState)
+    }
+  }, [hasUnsavedChanges])
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:375]: this component (the `CodeEditor`) is already really big, can you put this code in the parent component? Also can you make a dedicated hook e.g. `useWarnUserForUnsavedChanges(hasUnsavedChanges)`

**Comment by Satvik1769** [comment]
Ok will fix the issues

File: src/components/CodeEditor.tsx (hunk 2 above)
Comment (Satvik1769) [reply src/components/CodeEditor.tsx]: i created  a hook for the functionality now we just have to call it. it is in hooks folder with name useWarnUserForUnsavedChanges. I also reverted this code to previous state i.e.
const match = line.match(/@tsci\/[\w\-.]+/)

I hope there is no more conflicts and hope it helps.

File: src/components/CodeEditor.tsx [hunk 3]
```diff
@@ -323,6 +328,10 @@ export const CodeEditor = ({
     }
   }, [!isStreaming, currentFile, code !== ""])
 
+  useWarnUserForUnsavedChanges({
+    hasUnsavedChanges: hasUnsavedChanges ?? false,
+  })
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:333]: why does this have to be in the CodeEditor, can you move it to CodeAndPreview? This file is already very large and I don't see why this functionality has to be in here

File: src/hooks/useWarnUserForUnsavedChanges.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    const handlePopState = (event: PopStateEvent) => {
+      if (hasUnsavedChanges) {
+        const userConfirmed = window.confirm(
+          "Are you sure you want to go back? Unsaved changes may be lost.",
+        )
+        if (!userConfirmed) {
+          // Push the state again to stay on the same page
+          window.history.pushState(null, "", window.location.href)
```
Comment (seveibar) [inline src/hooks/useWarnUserForUnsavedChanges.ts:22]: do we have to push state to stay on the same page? Can we just not do anything and block the action?

File: src/hooks/useWarnUserForUnsavedChanges.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      }
+    }
+
+    // Attach event listeners
+    window.addEventListener("beforeunload", handleBeforeUnload)
+    window.addEventListener("popstate", handlePopState)
+
+    // Push the initial state only once, when component mounts
+    if (hasUnsavedChanges) {
+      window.history.pushState(null, "", window.location.href)
```
Comment (seveibar) [inline src/hooks/useWarnUserForUnsavedChanges.ts:33]: why do we need to push state here?

File: src/components/CodeEditor.tsx (hunk 3 above)
Comment (Satvik1769) [reply src/components/CodeEditor.tsx:333]: Yes it can also be moved in CodeAndPreview and it will work fine.

File: src/hooks/useWarnUserForUnsavedChanges.ts (hunk 4 above)
Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: state is necessary here this solution works by storing the current url in the state and if user clicks on cancel instead of sending it to previous page it send it to the current page

Comment (seveibar) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: What would happen if you removed the pushState?

Comment (seveibar) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: if i understand your code correctly, `preventDefault` stops the page from changing, so I'm not sure why you need to pushState

Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: It basically stores the current url (editor page) and only allows when cancelled. Is there a problem with storing the data in the state?

Comment (seveibar) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: Generally we only introduce state when absolutely necessary. If it's not necessary, we shouldn't do it.

I'm not convinced it's necessary. You only need pushState to change to a new page, and we're cancelling a navigation, so there should never be a need to pushState

File: src/hooks/useWarnUserForUnsavedChanges.ts (hunk 5 above)
Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:33]: we push the states two times:
1. At the starting of mounting if there is some unsaved changes
2. if person goes black but cancels then we repush the state to prevent moving back

File: src/hooks/useWarnUserForUnsavedChanges.ts (hunk 4 above)
Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: the thing is preventDefault will work for case like reloading but in case of user going back it will not work

Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: we can try to do this with pop state functionality but it doesn't block the action rather using it we can push state to restore current page

Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: I believe to add this functionality we will need to have states involved. However, if you have better method in mind I can make it .

Comment (seveibar) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: let's remove the pushStates, if we only work for reloading that's OK for now.

Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: as you wish but i will just comment out that part just in case you want to update

Comment (seveibar) [reply src/hooks/useWarnUserForUnsavedChanges.ts:22]: let's just remove it, i'd rather not have ghost code, people can come back and look at this PR.

Also make sure to revert the changes to files where things were accidentally changed. I'll leave some comments that indicate where that happened. There should be no changes to `CodeEditor` (IIUC), you can run `git checkout origin/main -- path/to/file` to revert a file

File: src/components/CodeEditor.tsx [hunk 6]
```diff
@@ -128,9 +128,13 @@ export const CodeEditor = ({
             .replace(registryPrefixes[2], "")
           const packageName = fullPackageName.split("/")[0].replace(/\./, "/")
           const pathInPackage = fullPackageName.split("/").slice(1).join("/")
-          const jsdelivrPath = `${packageName}${pathInPackage ? `/${pathInPackage}` : ""}`
+          const jsdelivrPath = `${packageName}${
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:131]: revert unnecessary change here

File: src/components/CodeEditor.tsx [hunk 7]
```diff
@@ -128,9 +128,13 @@ export const CodeEditor = ({
...
+            pathInPackage ? `/${pathInPackage}` : ""
+          }`
           return fetch(
-            `${apiUrl}/snippets/download?jsdelivr_resolve=${input.includes("/resolve/")}&jsdelivr_path=${encodeURIComponent(jsdelivrPath)}`,
+            `${apiUrl}/snippets/download?jsdelivr_resolve=${input.includes(
+              "/resolve/",
+            )}&jsdelivr_path=${encodeURIComponent(jsdelivrPath)}`,
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:137]: revert unnecessary change here

File: src/components/CodeEditor.tsx [hunk 8]
```diff
@@ -263,7 +267,7 @@ export const CodeEditor = ({
                 for (let pos = from; pos < to; ) {
                   const line = view.state.doc.lineAt(pos)
                   const lineText = line.text
-                  const matches = lineText.matchAll(/@tsci\/[\w\-.]+/g)
+                  const matches = lineText.matchAll(/@tsci\/[\w.]+/g)
```
Comment (seveibar) [inline src/components/CodeEditor.tsx:270]: this would introduce a bug, revert

File: src/hooks/useWarnUserForUnsavedChanges.ts (hunk 4 above)
Comment (Satvik1769) [reply src/hooks/useWarnUserForUnsavedChanges.ts]: ok will do

File: src/components/CodeEditor.tsx (hunk 8 above)
Comment (Satvik1769) [reply src/components/CodeEditor.tsx:270]: I am sorry for all this trouble i am thinking to shut this pr and create a cleaner pr that is easier to manage for you and ofcourse myself

Comment (seveibar) [reply src/components/CodeEditor.tsx:270]: yes that's always ok!

Comment (seveibar) [reply src/components/CodeEditor.tsx:270]: and no problem- my messages are short because i often type them from my phone, not because i'm angry 😁
